### PR TITLE
feat: Support API server timeout config

### LINKED ISSUE #2147: API server SLOs
- [ ]  `max-latency` & `timeout`
    - [x] api server timeout
    - [x]  provide both max-latency and timeout in BentoServer config
    - [x]  default `max-latency`:  `10s`
    - [ ]  default `timeout = 1.5 * max-latency`
    - [ ]  assert that `max-latency` < `timeout`
    - [ ]  provide `bentoml serve` CLI arg for `--max-latency`
    - [ ]  target: 90% < max-latency
- [ ]  `--max-request-size` ?
    - [ ]  provide BentoServer config
    - [ ]  default to `10MB`

### PR DESCRIPTION
**Pull request by frostming**
Signed-off-by: Frost Ming <me@frostming.com>

## What does this PR address?

This PR implements the timeout config for API server, it takes effect on the server side via a new middleware.

Fixes #2147

## Before submitting:

- [x] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [x] Does the code follow BentoML's code style, both `make format` and `make lint` script have passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [x] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [x] Did you write tests to cover your changes?

### CODE REVIEW

**Review by sauyon — APPROVED** [review]
LGTM!

File: tests/e2e/bento_server_http/tests/test_serve.py [hunk 1]
```diff
@@ -88,3 +92,17 @@ def test_serve_from_svc():
...
+    config_file = os.path.abspath("configs/timeout.yml")
+    env = os.environ.copy()
+    env.update(BENTOML_CONFIG=config_file)
+
+    with server.start(env=env) as client:
+        with pytest.raises(
+            BentoMLException,
+            match="504: b'Not able to process the request in 1 seconds'",
+        ):
+            client.echo_delay()
```
Comment (aarnphm) [inline tests/e2e/bento_server_http/tests/test_serve.py:108]: Let's update the config.yml to `server_config_file` fixture under this conftest

**Review by aarnphm — COMMENTED** [review]
one tiny comment.

**Review by aarnphm — COMMENTED** [review]
one tiny comment.

File: tests/e2e/bento_server_http/tests/test_serve.py (hunk 1 above)
Comment (frostming) [reply tests/e2e/bento_server_http/tests/test_serve.py:108]: The fixture is a matrix with other config files not related to the test. Should we update it?

Comment (aarnphm) [reply tests/e2e/bento_server_http/tests/test_serve.py:108]: I think it is fine just to do what you are doing.

**Pull request merged**